gobblin-metrics/src/main/java/gobblin/metrics/event/MultiTimingEvent.java

```diff
@@ -0,0 +1,143 @@
+ * starts a new timer for the new stage. A stage can be ended and timed without starting a new stage by calling
+ * {@link #endStage()}.
+ * When calling {@link #submit()}, the current stage will be ended, and a single event will be send containing the timing
+ * for each stage. Calling {@link #close()} will also submit the event.
+ * Calling {@link #submit()} or {@link #close()} when the event has already been called will throw an exception.
+ *
+ */
+public class MultiTimingEvent implements Closeable {
+
+  public static final String MULTI_TIMING_EVENT = "multiTimingEvent";
+
+  private final String name;
+  private final EventSubmitter submitter;
+  private final List<Stage> timings;
+  private boolean submitted;
+  private final boolean reportAsMetrics;
+  private String currentStage;
+  private long currentStageStart;
+
+  @Data
+  private class Stage {
+    private final String name;
+    private final long duration;
+  }
+
+  /**
+   * @param submitter {@link EventSubmitter} used to send the event.
+   * @param name Name of the event to be submitted.
+   * @param reportAsMetrics if true, will also report the timing of each stage to a timer with name <namespace>.<name>.<stage>
+   */
+  public MultiTimingEvent(EventSubmitter submitter, String name, boolean reportAsMetrics) {
+    this.name = name;
+    this.submitter = submitter;
+    this.timings = Lists.newArrayList();
+    this.submitted = false;
+    this.reportAsMetrics = reportAsMetrics;
+  }
+
+  /**
+   * End the previous stage, record the time spent in that stage, and start the timer for a new stage.
+   * @param name name of the new stage.
+   * @throws IOException
+   */
+  public void nextStage(String name) throws IOException {
+    endStage();
+    this.currentStage = name;
+    this.currentStageStart = System.currentTimeMillis();
+  }
+
+  /**
+   * End the previous stage and record the time spent in that stage.
+   */
+  public void endStage() {
+    if (this.currentStage != null) {
+      long time = System.currentTimeMillis() - this.currentStageStart;
+      this.timings.add(new Stage(this.currentStage, time));
+      if (reportAsMetrics && submitter.getMetricContext().isPresent()) {
+        String timerName = submitter.getNamespace() + "." + name + "." + this.currentStage;
+        submitter.getMetricContext().get().timer(timerName).update(time, TimeUnit.MILLISECONDS);
+      }
+    }
+    this.currentStage = null;
+  }
+
+  /**
+   * See {@link #submit(Map)}. Sends no additional metadata.
+   * @throws IOException
+   */
+  public void submit() throws IOException {
+    submit(Maps.<String, String>newHashMap());
+  }
+
+  /**
+   * Ends the current stage and submits the event containing the timings of each event.
+   * @param additionalMetadata additional metadata to include in the event.
+   * @throws IOException
+   */
+  public void submit(Map<String, String> additionalMetadata) throws IOException {
+
+    if (this.submitted) {
```

Comment:
MultiTimingEvent seems thread unsafe during close(), can we call it out in java docs?